### DISCUSSION TITLE
Feature Request: ability for admin to inject js into all Open-WebUI (adding shortcuts, better UI, etc)

### DISCUSSION THREAD

**Discussion by thiswillbeyourgithub** [open-webui/open-webui#5428] [Ideas]
The initial idea was to add keyboard shortcuts and make them customizable (by the admin, not the user). But to bring maximum extensibility for the admin and lower the burden on developpers I'm suggesting this somewhat larger and easier feature:

- In the admin settings, add  a new tab "js" containing a text field that will be injected into openwebui.
- This would make it trivial to add shortcuts, modify the UI, etc. It would give considerable control to the admin, even able to break it if they want too.
- Also needed would be a toggle switch at the top to disable the injection.
- As well as an env variable to disable the injection on startup, just as a last resort solution if something goes wrong.

An alternative would be to not modify the UI and simply inject the content of a local js file if present and specified by an env variable.

Here's a non exhaustive list of issue and discussions that would be solveable without bothering the devs if that was implemented:
- #3172 
- #2847
- #2869 
- #4426 
- #4754 
- #3158 
- #2846 
- #1414 
- #4682 
- #4422
- #5241
- #5147 
- #3960 
- #4074 
- #2698 
- #1132 
Probably more but basically anything that could be solved by an UI change can possibly be helped by this.

I'm aware there are ways to inject js into a page via browser extensions, but that won't work on mobile, work computers, my granny's computer etc.

It would also make Open-WebUI much able to customize the UI for people with accessibility issues, or tighten security by hiding some features from non admin users etc.

What do you think?

**Comment by tjbck** [comment]
My stance here is if they want to make any modifications they should fork the repo, however, I’d like to hear what other people have to say and potentially change my mind.

**Reply by alanxmay** [reply]
I love keyboard shortcut in VSCode, mapping command to key binding is really convenient.

**Reply by alanxmay** [reply]
Now the shortcut keys supported conflict with other global shortcut keys I have set.😢

**Reply by thomashkolbe** [reply]
I would like to be able to easily add a few lines of HTML and a small JS script to the Open Web-UI window, without requiring to fork the entire project. The code to be inserted should help in establishing a link between the Open Web-UI application and another web-application in a connected window. I am currently exploring a simple way to add a CesiumJS world browser (like Google Earth, check it here: https://sandcastle.cesium.com/) and then use Tools within Open Web-UI that query a spatial database (running in PostGIS) and then to show the results on the 3D Web Globe. In order to implement this, I need a possibility to launch some JS action from within a tool (that is running under Python). Some Tool functions should also query the virtual 3D globe, like "give me the list of selected 3D objects in the viewer".

**Comment by gkjpettet** [comment]
What I need is a way to inject a custom syntax highlighter into each web page since my work with LLMs is coding in a proprietary language that `highlight.js` doesn't support out of the box. I have already written a language definition for it but I need a way to inject it into the page.

**Comment by MatthiasLubbertsen** [comment]
where is this on the roadmap? would love to use this feature if it gets implemented.

**Comment by Classic298** [comment]
i think, this is actually already possible by moving custom js into the static folder in the custom area

**Comment by MatthiasLubbertsen** [comment]
What would be the process to do that? I would prefer not to build an image with custom code.

**Reply by Classic298** [reply]
you dont need to build a custom image, just move files into the correct folder inside your docker

**Reply by MatthiasLubbertsen** [reply]
Hmmm, I dont quite get it. Where should I move it, and how do I add that file to the frontend?
